File: Examples/C/Examples/Demux.c
#include "../Helpers.h"
#include <stdio.h>
#include "Ximu3.h"

static void PrintResult(const XIMU3_Result result);

void Demux()
{
    // Search for connection
    const XIMU3_Devices devices = XIMU3_port_scanner_scan_filter(XIMU3_PortTypeUsb);

    if (devices.length == 0)
    {
        printf("No USB connections available\n");
        return;
    }
    printf("Found %s %s\n", devices.array[0].device_name, devices.array[0].serial_number);

    // Open connection
    XIMU3_Connection* const connection = XIMU3_connection_new_usb(devices.array[0].usb_connection_info);

    XIMU3_devices_free(devices);

    if (XIMU3_connection_open(connection) != XIMU3_ResultOk)
    {
        printf("Unable to open connection\n");
        XIMU3_connection_free(connection);
        return;
    }

    // Demux
    const uint8_t channels[] = { 0x41, 0x42, 0x43 };
    const int numberOfChannels = sizeof(channels) / sizeof(channels[0]);

    XIMU3_Demux* const demux = XIMU3_demux_new(connection, channels, numberOfChannels);

    const XIMU3_Result result = XIMU3_demux_get_result(demux);

    PrintResult(result);

    if (result == XIMU3_ResultOk)
    {
        Wait(60);
    }

    XIMU3_demux_free(demux);

    // Close connection
    XIMU3_connection_close(connection);
    XIMU3_connection_free(connection);
}

static void PrintResult(const XIMU3_Result result)
{
    switch (result)
    {
        case XIMU3_ResultOk:
            printf("OK\n");
            break;
        case XIMU3_ResultError:
            printf("Error\n");
            break;
    }
    // printf("%s\n", XIMU3_result_to_string(result)); // alternative to above
}

File: Examples/Rust/src/examples/demux.rs
use ximu3::connection::*;
use ximu3::demux::*;
use ximu3::port_scanner::*;

pub fn run() {
    // Search for connection
    let devices = PortScanner::scan_filter(PortType::Usb);

    if devices.is_empty() {
        println!("No USB connections available");
        return;
    }

    let device = devices.first().unwrap();

    println!("Found {} {}", device.device_name, device.serial_number);

    // Open connection
    let connection = Connection::new(&device.connection_info);

    if connection.open().is_err() {
        println!("Unable to open connection");
        return;
    }

    // Demux
    let demux = Demux::new(&connection, vec![0x41, 0x42, 0x43]);

    print_result(&demux);

    if demux.is_ok() {
        std::thread::sleep(std::time::Duration::from_secs(60));
    }

    drop(demux);

    // Close connection
    connection.close();
}

fn print_result<T>(result: &std::io::Result<T>) {
    match result {
        Ok(_) => println!("OK"),
        Err(_) => println!("Error"),
    }
}
